# Pre-lecture HW - Cayden Wong 2024/09/08

### 1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

Session Link for Q1: https://chatgpt.com/share/1306abdd-53ed-48bb-a048-c6e2bd0829b4

In [10]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Count missing values
total_missing = df.isnull().sum().sum()
missing_per_column = df.isnull().sum()
missing_info = pd.DataFrame({
    'Missing Values': missing_per_column,
    'Percentage': df.isnull().mean() * 100
})

print(f"Total missing values in the dataset: {total_missing}")
print("\nMissing values per column:")
print(missing_per_column)
print("\nMissing values information:")
print(missing_info)


Total missing values in the dataset: 12

Missing values per column:
row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

Missing values information:
             Missing Values  Percentage
row_n                     0    0.000000
id                        1    0.255754
name                      0    0.000000
gender                    0    0.000000
species                   0    0.000000
birthday                  0    0.000000
personality               0    0.000000
song                     11    2.813299
phrase                    0    0.000000
full_id                   0    0.000000
url                       0    0.000000


##### Interaction Summary for Q1
Here's a summary of our interaction:

Initial Request:

You asked for a way to count missing values in a dataset from a provided URL using pandas.
Solution Provided:

I provided a step-by-step guide on how to load the dataset and count the missing values.
The steps included loading the dataset, calculating total missing values, missing values per column, and a detailed view with percentages.
Condensation Request:

You asked for the code to be condensed into a single block.
Condensed Code:

I provided a single block of code that accomplishes the tasks of loading the dataset, counting missing values, and displaying the results.
If you need further assistance or have any more questions, feel free to ask!

### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then
- use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,
- write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset

Session Link for Q2 + Q3: https://chatgpt.com/share/b1d6ee3a-15ce-4074-9282-b548bc7cf547

In [11]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Get the number of rows and columns
rows, columns = df.shape

print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 391
Number of columns: 11


**Observations**:
A row within a data set or a single point of information within a collection of data points. Each observation usually carries information about different variables / properties related to the observation.
Examples include: Information about people, objects, events, and places

**Variables**:
A property / charactristic that each observation may or may not have.
Examples include: someones age / height / weight / demographic, time / place of an event, population / GDP / inflation of a country.

### 3) Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

Session Link for Q2 + Q3: https://chatgpt.com/share/b1d6ee3a-15ce-4074-9282-b548bc7cf547

In [12]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Print column names
print("=== Column Names ===")
for i, col in enumerate(df.columns):
    print(f"{i + 1}. {col}")

# Print dataset info
print("\n=== Dataset Info ===")
df_info = df.info()

# Print numerical summary
print("\n=== Summary Statistics (Numerical Columns) ===")
numerical_summary = df.describe()
print(numerical_summary)

# Print categorical summary
print("\n=== Summary Statistics (Categorical Columns) ===")
categorical_summary = df.describe(include='object')
print(categorical_summary)

# Show first few rows for preview
print("\n=== First 5 Rows of Data ===")
data_preview = df.head()
print(data_preview)


=== Column Names ===
1. row_n
2. id
3. name
4. gender
5. species
6. birthday
7. personality
8. song
9. phrase
10. full_id
11. url

=== Dataset Info ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB

=== Summary Statistics (Numerical Columns) ===
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50

##### Interaction Summary for Q2 + Q3
Sure! Here’s a summary of our interaction:

1. **Dataset Overview**:
   - You are working with a dataset available at [this URL](https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv).
   - You wanted to know how to determine the number of rows and columns in the dataset using pandas.

2. **Determining Rows and Columns**:
   - You can use the `shape` attribute of a pandas DataFrame to get the number of rows and columns.
   - Example code:
     ```python
     import pandas as pd

     # Load the dataset
     url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
     df = pd.read_csv(url)

     # Get the number of rows and columns
     rows, columns = df.shape

     print(f"Number of rows: {rows}")
     print(f"Number of columns: {columns}")
     ```

3. **Understanding Observations and Variables**:
   - **Observations**: Individual data points or rows in the dataset (e.g., each villager).
   - **Variables**: Attributes or columns in the dataset that describe observations (e.g., name, age, species).

4. **Extracting Information About Columns**:
   - To get a summary of columns, you can use methods like:
     - `df.info()` for a concise summary.
     - `df.describe()` for numerical columns' statistics.
     - `df.describe(include='object')` for categorical columns' statistics.
     - `df.columns` to list column names.
     - `df.head()` to preview the first few rows.

5. **Formatting Output**:
   - You can format the output to be more readable by clearly labeling sections and providing structured information.
   - Example of structured output:
     ```python
     print("=== Column Names ===")
     for i, col in enumerate(df.columns):
         print(f"{i + 1}. {col}")

     print("\n=== Dataset Info ===")
     df_info = df.info()

     print("\n=== Summary Statistics (Numerical Columns) ===")
     numerical_summary = df.describe()
     print(numerical_summary)

     print("\n=== Summary Statistics (Categorical Columns) ===")
     categorical_summary = df.describe(include='object')
     print(categorical_summary)

     print("\n=== First 5 Rows of Data ===")
     data_preview = df.head()
     print(data_preview)
     ```

6. **Understanding Numerical Summary**:
   - The numerical summary provides key statistics for numerical columns:
     - **Count**: Number of non-null values.
     - **Mean**: Average value.
     - **Standard Deviation (std)**: Measure of spread.
     - **Min**: Minimum value.
     - **25%**, **50%** (median), **75%**: Quartiles.
     - **Max**: Maximum value.
   - These statistics help assess data completeness, distribution, and variability.

Feel free to ask if you need any further clarification or additional details!

### 4) If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

Session Link for Q4: https://chatgpt.com/share/fd9be579-b2d3-4bc0-9e31-3d70e9666d16

In [13]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

print(df.shape)

print(df.describe(include='all'))
# print(df.describe())

(891, 15)
          survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \

In [14]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [15]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**Why is the number of variables different that what is described in `.shape` using `.describe()`?**
When using `.describe()` without any additional parameters, the function only returns data about variables that contain only integer values. This means columns that include information such as name, class, and embark_town are not included when using `.describe()` leading to a lower percived number of variables. This can be fixed to include all types by adding the following additional argument: `.describe(include='all')`.

**Why is the count number different between variables?**
The count number is different to the number described by `.shape` as it does not count missing values contained within the dataset. For example, the variable survived returns  a count of `891` (the same as described by `.shape`) while age only returns a count of `714`. This likely indicates that there is missing data about the age of some of the observations within the dataset, most likely information that was unobtainable because the passenger may have not survived.

### 5) Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

Session Link for Q4 + Q5: https://chatgpt.com/share/fd9be579-b2d3-4bc0-9e31-3d70e9666d16

**Attribute** (Example: `df.shape`)
An attribute is an intrinsic property of some object that stores information or data about that object (in the case of `.shape`, it is a property describing the dimensions of a pandas DataFrame object). It is not a function call and does not need to be called with any parameters.

**Method** (Example: `df.describe()`)
A method is an property of some object that performs and operation / returns information about said object (for .describe(), this is a method of the pandas DataFrame object). Since it is a function call, it can take in additional paramaters within the `()`.

##### Interaction summary for Q4 + Q5
In this conversation:

You were working with a dataset from the Titanic (https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv) using pandas and wanted to know how to make df.describe() output summary statistics for all data types, not just numerical ones. I suggested using the include='all' parameter in df.describe() to include all data types (numerical, categorical, etc.).

You asked why the top row in df.describe(include='all') sometimes reports NaN as its value. I explained that this can happen when:

The column contains mostly or entirely NaN values.
There are mixed data types in the column.
NaN is the most frequent value in that column.
You asked about the difference between an attribute and a method. I explained that an attribute holds data or properties of an object (like .shape), while a method is a function that performs an action (like .describe()), requiring parentheses to invoke it.

Let me know if you'd like more details or clarification on any part!

# Post-lecture HW - Cayden Wong 2024/09/12

### 6) The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics

Session Link for Q6: https://chatgpt.com/share/270d3b1c-5d2c-47c1-8053-57d4e654244c

In [16]:
df.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

### Terminology

**Count**: The number of null data points for the given variable.  
**Mean**: The average of all the data points ($\bar x = \frac{1}{n}\sum^n_{i=1}x_i$).  
**Standard Deviation (std)**: A figure representing the variance between the mean and the individual data points. Low values represent little varience, larger values represent larger varience.  
**Minimum (min)**: The smallest element within a given variable.  
**First Quartile (25%)**: The value which 25% of the data points fall below.  
**Median (50%)**: The value which 50% of the data points fall below.  
**Third Quartile (75%)**: The value which 75% of data points fall below.  
**Maximum (max)**: The argest element within a given variable.

##### Interaction Summary for Q6
Here’s a summary of our interaction:

1. **Statistical Terms**:
   - You asked for an explanation of statistical terms like count, mean, standard deviation (std), minimum (min), quartiles (25%, 50%, 75%), and maximum (max). I provided definitions of each term and explained their use in describing the distribution of data.

2. **Count and Missing Data**:
   - You inquired if the "count" includes missing data, and I explained that the count typically excludes missing values in most statistical calculations.

3. **Handling Missing Data in `df.describe()`**:
   - You asked how the `df.describe()` function in pandas handles missing data. I clarified that pandas ignores `NaN` values when calculating summary statistics like mean, standard deviation, and quartiles. Only non-missing values are included in these calculations.

Let me know if you need further clarification or details!

### 7) Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

Session Link for Q7: https://chatgpt.com/share/9da81e36-fc99-465b-9a1e-ce9bdfc1c566

##### Provide an example of a "use case" in which using `df.dropna()` might be peferred over using del `df['col']`

One might prefer to use `df.dropna()` over `df['col']` if their data set contains observations that are missing data-points for several different varaibles. Deleting these observations using `df.dropna()` may be best as their incompleteness may skew / portray the general data in an inaccurate manner. Attempting to remove the few missing data-points using `df['col']` would remove many data-points that do exists in most other observations.

##### Provide an example of "the opposite use case" in which using del `df['col']` might be preferred over using `df.dropna()`

`df['col']`s use may be preferred if many observations contain missing data-points for one specific variable. Since `df.dropna()` would remove all of the observations with missing data, many data points that would provide insight on other variables within the dataset would be lost. Contrarily, `df['col']` would discard the variable containg mostly missing information without losing data points about other variables from the rest of each observation.

##### Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important

Since `df.dropna()` removes all observations with missing data, if we are using `del df['col']` to remove a column that contains many points of missing data, we would want to use it before `df.dropna()`. Consider the scenario where many observations only contain a singular missing data-point all under a single variable. Using `df.dropna()` first would result in all of those observations being removed from the dataset. However, with use of `del df['col']`, we can remove the variable with many missing data-points before `df.dropna()` is run thus preserving the observations which would have otherwise been deleted. 

##### Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

In [17]:
df_copy = df.copy()

# df_nona = df.dropna() This would reduce observations to < 200
del df_copy['deck'] # Remove 'deck' col as it contains many missing values
df_cleaned = df_copy.dropna()
df_cleaned.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   712.000000  712.000000   712  712.000000  712.000000  712.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   453         NaN         NaN         NaN   
mean      0.404494    2.240169   NaN   29.642093    0.514045    0.432584   
std       0.491139    0.836854   NaN   14.492933    0.930692    0.854181   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    1.000000   NaN   20.000000    0.000000    0.000000   
50%       0.000000    2.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    1.000000   
max       1.000000    3.000000   NaN   80.000000    5.000000    6.000000   

              fare embarked  class  who adult_male  embark_town alive alone  
count   712.000000      712    712  712        712          712   712   712  
unique         NaN        3      3    3          2            3     2     2  
top            NaN        S  Third  man       True  Southampton    no  True  
freq           NaN      554    355  413        413          554   424   402  
mean     34.567251      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
std      52.938648      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
min       0.000000      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
25%       8.050000      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
50%      15.645850      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
75%      33.000000      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
max     512.329200      NaN    NaN  NaN        NaN          NaN   NaN   NaN

In the dataset above, I decided to remove the `'deck'` col from the dataset before using `.dropna()`. This is because over 600 data-points were missing from this column. If `dropna()` were to be used first, 600 observations would be removed from the data. However, after removing the `'deck'` col, using `.dropna()` only removed around 200 observations instead, thus preserving a larger amount of data.

##### Interaction summary for Q7:
Here’s a summary of our interaction:

1. **df.dropna() vs. del df['col']**:
   - You asked about the differences between `df.dropna()` and `del df['col']` in pandas. I explained that `df.dropna()` is used for removing rows or columns with missing values, while `del df['col']` is for deleting a specific column. Key differences include their usage, return values, and how they modify data.
   
2. **Dataset with missing data points**:
   - You asked for datasets with missing values in a single variable. I suggested the Titanic and Ames Housing datasets, highlighting that *Age* in Titanic and *PoolQC* in Ames Housing often have missing values.

3. **Finding a housing price dataset**:
   - You requested a housing price dataset with missing pool values in CSV format. After searching, I provided links, including the Ames Housing dataset from Kaggle, which has *PoolQC* as a column with mostly missing values.

I can continue assisting you if you'd like further information!

### 8) Give brief explanations in your own words for any requested answers to the questions below

Session link for Q8: https://chatgpt.com/share/c241f801-4aea-4870-bf54-7ddd73a3fedf

In [2]:
import pandas as pd
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(titanic_url)

In [10]:
titanic_df.describe(include="all")

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

In [4]:
titanic_df.groupby("survived").describe()

pclass                                                 age  \
          count      mean       std  min  25%  50%  75%  max  count   
survived                                                              
0         549.0  2.531876  0.735805  1.0  2.0  3.0  3.0  3.0  424.0   
1         342.0  1.950292  0.863321  1.0  1.0  2.0  3.0  3.0  290.0   

                     ... parch        fare                             \
               mean  ...   75%  max  count       mean        std  min   
survived             ...                                                
0         30.626179  ...   0.0  6.0  549.0  22.117887  31.388207  0.0   
1         28.343690  ...   1.0  5.0  342.0  48.395408  66.596998  0.0   

                                         
              25%   50%   75%       max  
survived                                 
0          7.8542  10.5  26.0  263.0000  
1         12.4750  26.0  57.0  512.3292  

[2 rows x 40 columns]

In [8]:
titanic_df.groupby("pclass")["fare"].describe()

count       mean        std  min       25%      50%   75%       max
pclass                                                                     
1       216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
2       184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
3       491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

In [11]:
titanic_df.groupby("sex")["pclass"].describe()

count      mean      std  min  25%  50%  75%  max
sex                                                      
female  314.0  2.159236  0.85729  1.0  1.0  2.0  3.0  3.0
male    577.0  2.389948  0.81358  1.0  2.0  3.0  3.0  3.0

In [7]:
titanic_df.groupby("pclass")["age"].describe()

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [14]:
titanic_df.groupby("sex")[["age", "survived"]].describe()

age                                                     survived  \
        count       mean        std   min   25%   50%   75%   max    count   
sex                                                                          
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0    314.0   
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0    577.0   

                                                     
            mean       std  min  25%  50%  75%  max  
sex                                                  
female  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0  
male    0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

In [9]:
titanic_df.groupby("pclass")[["parch", "sibsp"]].describe()

parch                                               sibsp            \
        count      mean       std  min  25%  50%  75%  max  count      mean   
pclass                                                                        
1       216.0  0.356481  0.693997  0.0  0.0  0.0  0.0  4.0  216.0  0.416667   
2       184.0  0.380435  0.690963  0.0  0.0  0.0  1.0  3.0  184.0  0.402174   
3       491.0  0.393075  0.888861  0.0  0.0  0.0  0.0  6.0  491.0  0.615071   

                                           
             std  min  25%  50%  75%  max  
pclass                                     
1       0.611898  0.0  0.0  0.0  1.0  3.0  
2       0.601633  0.0  0.0  0.0  1.0  3.0  
3       1.374883  0.0  0.0  0.0  1.0  8.0

##### Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like `df.groupby("col1")["col2"].describe()`?

The count values within `df.describe()` describe the number of observations that do not have a null value each variable.
On the contary, when using the `.groupby('col1')` method, we restrict the dataframe to only contain information regarding observations that have a defined value for the `'col1'` variable, re-organizing them by their value and essentially describing the frequency of each non-null value within `'col1'`. This allows us to draw further conclusions about the relationship between values in `'col1'` to values in `'col2'`. However, this does not allow us to conclude anything about the missing data and can potentially mislead our observations by not including information about the observations with missing values.

##### Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT


**Forget to include import pandas as pd in your code**
```python
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.cvs"
titanic_df = pd.read_csv(titanic_url)
titanic_df.groupby("survived")["parch"].describe()
```
Both ChatGPT and searching google took around the same amount of time to identify and fix the error.  
ChatGPT: Pointed to a simpler more common source of error as the culprit, (missing import), did not identify any other potential sources of error.  
Google: Pointed to StackOverflow as the top result which linked to an thread where someone was not missing the inport statement and was still having problems.  

**Mistype "titanic.csv" as "titanics.csv"**
```python
import pandas as pd
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.cvs"
titanic_df = pd.read_csv(titanic_url)
titanic_df.groupby("survived")["parch"].describe()
```
In this case, ChatGPT was much better at identifing the error than google, it isn't clear exactly what to search up in google for the potential cause of the error.  
ChatGPT: Immediately identified the two spelling mistakes in the URL.  
Google: Was not able to easily find a source that identifies the error.  

**Try to use a dataframe before it's been assigned into the variable**
```python
import pandas as pd
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(titanic_url)
titanicdf.groupby("survived")["parch"].describe()
```
Again, it was easier to find a solution using ChatGPT as it was able to point to exactly where the error was. Google on the other hand shows solutions to many different unrelated issues that are not specific to a simple error like this.  
ChatGPT: Immediately identified the spelling mistake.  
Google: Pointed to many different instances of people encountering NameErrors but from reasons other than misspelling a variable name.  

**Forget one of the parentheses somewhere the code**
```python
import pandas as pd
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(titanic_url)
titanic_df.groupby("survived"["parch"].describe()
```
For this case, ChatGPT gave a faster precise and specific response. Google leads to other posts with similar but unrelated issues.  
ChatGPT: Correctly fixed the error and provided working code.  
Google: Points to a post where a user encounters `SyntaxError: incomplete input` for a different reason.  

**Mistype one of the names of the chained functions with the code**
```python
import pandas as pd
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.readcsv(titanic_url)
titanic_df.group_by("survived")["parch"].describe()
```
Using Google may have allowed the user to identify the error in the same timeframe as ChatGPT. However, the user would need to do multiple searches per error, whereas ChatGPT corrected both.  
ChatGPT: Found and corrected both errors and provided working code in return.  
Google: Linked to a post where someone correctly wrote `read_csv` but was still getting the error, this may give the user a hint on the reason why the code errors.  

**Use a column name that's not in your data for the groupby and column selection**
```python
import pandas as pd
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(titanic_url)
titanic_df.groupby("name")["age"].describe()
```
It is unclear what to search for in Google to solve this issue, however, it does point to hints on what the issue is. ChatGPT immediately solves the isse and provides the user with feedback on what to fix.  
ChatGPT: Identified the problem and suggested other column names to try instead.  
Google: Pointed to a thread where the user got a missing column name KeyError had the column in their dataframe (their issue was with their csv file delimiter)  

**Forget to put the column name as a string in quotes for the groupby and column selection**
```python
import pandas as pd
titanic_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(titanic_url)
titanic_df.groupby(survived)[age].describe()
```
In this example, Google could not provide any information on why the code would not work as the issue is too elementary and case specific to find a general answer. However, ChatGPT identified the issue without a problem.  
ChatGPT: Identified the issue was with missing quotation marks and provided corrected code.  
Google: Does not really give any indication to why the code errors.  

##### Interaction Summary for Q8: 
Here's a summary of our entire interaction:

1. **Initial Request:** You mentioned you downloaded the Titanic dataset and wanted help understanding the code you were working with for analysis.
   
2. **Code Explanation:**
   - I explained how `df.groupby("col1")["col2"].describe()` works, which groups a DataFrame by `"col1"`, selects column `"col2"`, and provides descriptive statistics for each group.

3. **First Error (NameError: name 'pd' is not defined):**
   - This occurred because the `pandas` library wasn't imported. I suggested adding `import pandas as pd` at the top of your script.

4. **Second Error (HTTPError):**
   - The dataset URL had a typo (`"titanics.cvs"` instead of `"titanic.csv"`). I pointed out the correct URL.

5. **Third Error (NameError: name 'titanicdf' is not defined):**
   - The variable `titanic_df` was misspelled as `titanicdf`. I advised using the correct variable name consistently.

6. **Fourth Error (SyntaxError: incomplete input):**
   - The column selection syntax was incorrect. I provided the correct format: `titanic_df.groupby("survived")["parch"].describe()`.

7. **Fifth Error (AttributeError: module 'pandas' has no attribute 'readcsv'):**
   - There was a typo in the function name (`pd.readcsv`). I corrected it to `pd.read_csv()`.

8. **Sixth Error (KeyError: 'name'):**
   - The column `"name"` didn’t exist in the DataFrame. I suggested checking the available columns using `titanic_df.columns` to find the correct column names.

9. **Seventh Error (NameError: name 'survived' is not defined):**
   - The column names were referenced without quotes. I explained that column names must be strings (i.e., `"survived"` and `"age"`).

Each error was due to small typos or syntax mistakes, and I guided you through fixing them. Would you like any further clarifications or additional help with the analysis?

### 9) Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

Not yet, however, I will have by the time this HW assignment is submitted.